stablehlo/transforms/VhloLegalizeToStablehlo.cpp

```diff
@@ -464,6 +464,198 @@ SpecialResult convertSpecial(const OpConversionPattern<VhloOpTy>& pattern,
   return notSpecial();
 }
 
+//===----------------------------------------------------------------------===//
+// VHLO --> StableHLO attributes: 3) Default attributes.
+// Unlike StableHLO, VHLO doesn't have default attributes, so the corresponding
+// attributes are added explicitly during StableHLO --> VHLO conversion.
+// These attributes are removed here. (Strictly speaking, don't have to do this
+// but this makes eyeballing easier).
+//===----------------------------------------------------------------------===//
+
+bool isBoolean(Attribute vhloAttr, bool value) {
+  auto attr = vhloAttr.template dyn_cast_or_null<vhlo::BooleanV1Attr>();
+  return attr && attr.getValue() == value;
+}
+
+bool isEmptyArray(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::ArrayV1Attr>();
+  return attr && attr.getValue().empty();
+}
+
+bool isEmptyString(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::StringV1Attr>();
+  return attr && attr.getValue().empty();
+}
+
+bool isEmptyTensor(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<DenseElementsAttr>();
+  return attr && attr.getNumElements() == 0;
+}
+
+bool isEnum(Attribute vhloAttr, Attribute value) { return vhloAttr == value; }
+
+bool isInteger(Attribute vhloAttr, int64_t value) {
+  auto attr = vhloAttr.template dyn_cast_or_null<vhlo::IntegerV1Attr>();
+  return attr && attr.getValue().getSExtValue() == value;
+}
+
+bool isSplatArray(Attribute vhloAttr, Attribute splatValue) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::ArrayV1Attr>();
+  return attr && llvm::all_of(attr.getValue(), [&](Attribute attr) {
+           return attr == splatValue;
+         });
+}
+
+template <typename T>
+bool isSplatTensor(const ConversionPattern& pattern, Attribute vhloAttr,
+                   T splatValue) {
+  auto attr = convertGeneric(vhloAttr, pattern.getTypeConverter())
+                  .template dyn_cast_or_null<DenseElementsAttr>();
+  return attr && attr.isSplat() &&
+         attr.template getSplatValue<T>() == splatValue;
+}
+
+bool isString(Attribute vhloAttr, StringRef value) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::StringV1Attr>();
+  return attr && attr.getValue() == value;
+}
+
+// TODO: Also validate attributes before removing them.
```

Comment:
What does this mean? Is `dyn_cast_or_null` and compare to value not enough? Or is this more for integers/bool comparisons?

Reply:
Opened a ticket about that: https://github.com/openxla/stablehlo/issues/1232.